orttraining/orttraining/core/session/training_session.cc

```diff
@@ -916,11 +1040,211 @@ common::Status TrainingSession::Run(const RunOptions& run_options, IOBinding& io
   return InferenceSession::Run(run_options, io_binding);
 }
 
+void TrainingSession::CreateBatchVariables(
```

Comment:
i.e. batch = microbatch?  Should this be in a pipeline-specific file rather than the main TrainingSession?

Reply:
Yes, it creates micro-batch. In the following `RunWithPipeline`, we call
```c++
    // Add inputs and outputs to the binding.
    CreateBatchVariables(io_binding, *sub_io_binding.get(), i, num_steps);

    // Add proper events to the binding.
    CreatePipelineEvents(training_mode, i, stage_id, *sub_io_binding.get());
```
The pipeline schedule is a field of `TrainingSession`, so `CreatePipelineEvents` is a member function of `TrainingSession`. Then, it looks natural to put `CreateBatchVariables` in `TrainingSession`. In addition, the batch variables created are associated (allocated) by `TrainingSession` and this function is tightly coupled with `TrainingSession`.

---
In reply to: [527651152](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r527651152) [](ancestors = 527651152)